src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -29,20 +35,66 @@ public SolutionAssetProvider(HostWorkspaceServices services)
         {
-            // Complete RPC right away so the client can start reading from the stream.
-            // The fire-and forget task starts writing to the output stream and the client will read it until it reads all expected data.
-            _ = Task.Run(async () =>
+            var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
+            var serializer = _services.GetRequiredService<ISerializerService>();
+
+            SolutionAsset? singleAsset = null;
+            IReadOnlyDictionary<Checksum, SolutionAsset>? assetMap = null;
+
+            if (checksums.Length == 1)
             {
-                using var writer = new ObjectWriter(outputStream, leaveOpen: false, cancellationToken);
+                singleAsset = (await assetStorage.GetAssetAsync(scopeId, checksums[0], cancellationToken).ConfigureAwait(false)) ?? SolutionAsset.Null;
+            }
+            else
+            {
+                assetMap = await assetStorage.GetAssetsAsync(scopeId, checksums, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC pipe.
+            // Run two tasks - the first synchronously writes to a local pipe and the second asynchronosly transfers the data to the RPC pipe.
+            //
+            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but might result in more
+            // (non-contiguous) memory allocated for the underlying buffers. The amount of memory is bounded by the total size of the serialized assets.
+            var localPipe = new Pipe(RemoteHostAssetSerialization.PipeOptionsWithUnlimitedWriterBuffer);
 
-                var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
-                var serializer = _services.GetRequiredService<ISerializerService>();
-                await RemoteHostAssetSerialization.WriteDataAsync(writer, assetStorage, serializer, scopeId, checksums, cancellationToken).ConfigureAwait(false);
-            }, cancellationToken);
+            Task.Run(() =>
+            {
+                try
+                {
+                    using var stream = localPipe.Writer.AsStream(leaveOpen: false);
+                    using var writer = new ObjectWriter(stream, leaveOpen: false, cancellationToken);
+                    RemoteHostAssetSerialization.WriteData(writer, singleAsset, assetMap, serializer, scopeId, checksums, cancellationToken);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e, cancellationToken))
+                {
+                    // no-op
+                }
+            }, cancellationToken).Forget();
+
+            async Task CopyPipeData()
```

Comment:
can you place hte local functions after all the main body code.  it makes it much easier to read.  with htis, i had to intersperse different method scopes together to figure out what was going on.

Reply:
```suggestion
            async Task CopyPipeDataAsync()
```

Reply:
Sure, but there is only one local function. The other is lambda.